### PR TITLE
Probe TLS and credentials configuration feature.

### LINKED ISSUE #3766: Probe TLS and credentials configuration
**What is missing?**

Probes currently cannot configure TLS and/or authentication for the prober they use.

**Why do we need it?**

A prober could fairly easily be abused by an unauthorized user to DDoS a target, so it's rather important to secure the probing endpoint. As it currently stands though, this is not possible.

**Environment**

Any version of the prometheus operator that has Probe support.

**Anything else we need to know?**:

Without digging deeper, I think a lot of what we did for PodMonitors should be re-usable for this.

### PR DESCRIPTION
**Pull request by mikechengwei**
Fix #3766.
Add TLS and credentials configuration to probe configuration.

```release-note:feature
Probe TLS and credentials configuration feature.
```

### CODE REVIEW

File: pkg/prometheus/operator.go [hunk 1]
```diff
@@ -1834,6 +1834,10 @@ func (c *Operator) selectProbes(p *monitoringv1.Prometheus) (map[string]*monitor
 			)
 			continue
 		}
+		pnKey := fmt.Sprintf("probe/%s/%s/%d", probe.GetNamespace(), probe.GetName())
+		if err = store.AddBearerToken(ctx, probe.GetNamespace(), probe.Spec.BearerTokenSecret, pnKey); err != nil {
```
Comment (brancz) [inline pkg/prometheus/operator.go:1838]: I think this needs something similar to what's in the selectservicemonitor and podmonitor functions for TLS assets to also be selected

File: pkg/apis/monitoring/v1/types.go [hunk 2]
```diff
@@ -875,6 +875,15 @@ type ProbeSpec struct {
 	Interval string `json:"interval,omitempty"`
 	// Timeout for scraping metrics from the Prometheus exporter.
 	ScrapeTimeout string `json:"scrapeTimeout,omitempty"`
+	// TLS configuration to use when scraping the endpoint.
+	TLSConfig *PodMetricsEndpointTLSConfig `json:"tlsConfig,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:879]: for clarity, best to rename it `ProbeTLSConfig`?

File: pkg/prometheus/operator.go [hunk 3]
```diff
@@ -1835,6 +1835,14 @@ func (c *Operator) selectProbes(p *monitoringv1.Prometheus) (map[string]*monitor
...
 			continue
 		}
+		pnKey := fmt.Sprintf("probe/%s/%s", probe.GetNamespace(), probe.GetName())
+		if err = store.AddBearerToken(ctx, probe.GetNamespace(), probe.Spec.BearerTokenSecret, pnKey); err != nil {
+			break
+		}
+
+		if err = store.AddBasicAuth(ctx, probe.GetNamespace(), probe.Spec.BasicAuth, pnKey); err != nil {
+			break
+		}
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:1845]: We still need to call `store.AddTLSConfig()` if the TLS config isn't nil?

Comment (mikechengwei) [reply pkg/prometheus/operator.go:1845]: Yes , you're right.

**Comment by simonpasquier** [comment]
can you run `make generate` and commit the changes?

**Comment by mikechengwei** [comment]
@simonpasquier Hi , i have done your requirements and you can review this pr again.

File: pkg/apis/monitoring/v1/types.go [hunk 4]
```diff
@@ -883,6 +883,15 @@ type ProbeSpec struct {
 	Interval string `json:"interval,omitempty"`
 	// Timeout for scraping metrics from the Prometheus exporter.
 	ScrapeTimeout string `json:"scrapeTimeout,omitempty"`
+	// TLS configuration to use when scraping the endpoint.
+	TLSConfig *ProbeTLSConfig `json:"tlsConfig,omitempty"`
+	// Secret to mount to read bearer token for scraping targets. The secret
+	// needs to be in the same namespace as the pod monitor and accessible by
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:889]:
```suggestion
	// needs to be in the same namespace as the probe and accessible by
```

**Review by simonpasquier — COMMENTED** [review]
one small nit otherwise lgtm

**Comment by simonpasquier** [comment]
:+1:  you need to run `make generate` and commit the changes though

**Review by simonpasquier — APPROVED** [review]
thanks!

**Pull request merged**